pkg/ipam/ippool.go

```diff
@@ -0,0 +1,209 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package ipam
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+
+	"github.com/emirpasic/gods/maps/treemap"
+)
+
+type IPPool struct {
+	cidr      string
+	network   *net.IPNet
+	size      int
+	available *treemap.Map // int IP is the key, string IP is the value
+	sync.RWMutex
+}
+
+func NewIPPool(cidr string) (*IPPool, error) {
+	_, network, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+
+	ones, totalbits := network.Mask.Size()
+	size := int(math.Exp2(float64(totalbits-ones))) - 2 // don't count net and broadcast
+	if size < 2 {
+		return nil, fmt.Errorf("invalid prefix for CIDR %q", cidr)
+	}
+
+	pool := &IPPool{
+		cidr:      cidr,
+		network:   network,
+		size:      size,
+		available: treemap.NewWithIntComparator(),
+	}
+
+	startingIP := ipToInt(pool.network.IP) + 1
+
+	for i := 0; i < pool.size; i++ {
+		intIP := startingIP + i
+		ip := intToIP(intIP).String()
+		pool.available.Put(intIP, ip)
+	}
+
+	// TODO: RecordAvailability(cidr, size)
+
+	return pool, nil
+}
+
+func ipToInt(ip net.IP) int {
+	intIP := ip
+	if len(ip) == 16 {
+		intIP = ip[12:16]
+	}
+
+	return int(binary.BigEndian.Uint32(intIP))
+}
+
+func intToIP(ip int) net.IP {
+	netIP := make(net.IP, 4)
+	binary.BigEndian.PutUint32(netIP, uint32(ip))
+
+	return netIP
+}
+
+func StringIPToInt(stringIP string) int {
+	return ipToInt(net.ParseIP(stringIP))
+}
+
+func (p *IPPool) allocateOne() ([]string, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	iter := p.available.Iterator()
+	if iter.Last() {
+		p.available.Remove(iter.Key())
+		// TODO: RecordAllocateGlobalIP(p.cidr)
+
+		return []string{iter.Value().(string)}, nil
+	}
+
+	return nil, errors.New("insufficient IPs available for allocation")
+}
+
+func (p *IPPool) Allocate(num int) ([]string, error) {
+	if num < 0 {
```

Comment:
nitpick: Can't this be a switch case? Surprised linting didn't complain about it.

Reply:
done